### ISSUE TITLE
Feature request: "<vs cities [cityFilter]>"

### ISSUE THREAD

**Issue by SpacedOutChicken** [yairm210/Unciv#10124]
I'd like to see the `"+30% Strength when fighting City-State units and cities"` unique fully parameterized. It's almost possible to do this by doing `"[+30]% Strength <vs [City-State] units>"`, but that doesn't apply to cities, and the only conditional that does target cities does not distinguish between city types. I propose adding a conditional that does distinguish between city types. For lacking a better format, I propose doing it as `"<vs cities [cityFilter]>"`.

This conditional could give units a bonus for attacking city-state cities, or for attacking annexed cities, or holy cities, or coastal cities, and possibly other kinds of cities if given the right parameters.

**Comment by hackedpassword** [comment]
Readability is also for a very limited audience. Key is documentation, the pillar of comprehending the hieroglyphs we conjure in json configs.

Both proposals are readable to me, Trog's seems more limited, maybe missing the flexibility SpacedOutChicken is trying to achieve.

**Comment by SpacedOutChicken** [comment]
Maybe it's time to rethink city filters? Redo all the uniques to use `"<in [cityFilter] cities"`?

**Comment by SeventhM** [comment]
I'll need to double check, But last time I whined about cityFilters, I believe the big issue is how some city filters worked. I'm not sure if cb7bb51e18d98cdb4630d7231b21c2802cc51e11 also helped move towards allowing city filters to be in condtionals, but it is the reason why we, for instance, we has a specific `<in this city>` instead of a generic `<[cityFilter]>`

**Comment by SeventhM** [comment]
Also, double checking the wording on cityFilter, it definitely would need an overhaul to make sense in a conditional without looking crazy

**Comment by yairm210** [comment]
I'm game for redesigning city filters, I think we've learnt a lot about filtering in the last few years.

First off, we've learnt that Modders Prefer Tags. 
A tag system is one where you can add arbitrary tags to objects, and then filter by that tag - the way that unit uniques and promotions work for filtering units. What we may need to do is find a better way of inputting that data,  `[{Wounded} {Melee} {Water}] units` does work but it's a bit ugly.

The tags we can currently add to cities are:
- Capital
- {nationFilter}
-  Annexed
- Puppeted
- Holy
- Connected

The other filters don't really work with this system, because they go on the *end* of the text.
 `in which the majority religion is a major religion` sure is a mouthful, I can't see a good way to fit that in the system.

Second, we've learnt that modders like meta-modifiers. 
So a building that adds a tag to a city would have a unique like "Makes this city a [tag] city", and then other buildings can use the "<in [cityFilter] cities>" - kind of like [Power in Civ IV](https://civilization.fandom.com/wiki/Power_(Civ4))
In fact Civ IV Power could be implemented entirely through this method - with conditionals on that unique to restrict coal plants only providing it when we have coal, etc

Since the new filter *cannot* replace the old one 100%, at least how I'm imagining it, we we have them side by side... So what to call the new one?  "newCityFilter"? "cityTagFilter"?

**Comment by yairm210** [comment]
```
"in this city",
"in all cities", -> "All"
"in all coastal cities", -> "Coastal"
"in capital", -> "Capital" (this would mean "in [Capital] cities" which is... eh])
"in all non-occupied cities", -> "Non-occupied"
"in all cities with a world wonder", -> "Wonder-containing" ???
"in all cities connected to capital", -> "Connected"
"in all cities with a garrison", -> "Garrisoned" ???
"in all cities in which the majority religion is a major religion",
"in all cities in which the majority religion is an enhanced religion",
"in non-enemy foreign cities",
"in foreign cities", 
"in annexed cities", -> "Annexed"
"in puppeted cities", -> "Puppeted"
"in holy cities", -> "Holy"
"in City-State cities", -> nationFilter already contains City-State
"in cities following this religion",
```

SOO 
The first item that arises is that cityFilter is currently *observer dependent* - what is "foreign" what is "enemy" - whereas most of the filters are *not*. A challenge we'll need to contend with.
The second is that the religion filters really much up everything. Yuck.

**Comment by yairm210** [comment]
Actually let's turn this into an RFC. @SpacedOutChicken @SeventhM What do you think of the following idea?

- ALL filters will obey "{conditionA} {conditionB}" (AND)
- ALL filters will accept "Non-[condition]" for negation (NOT)

So like "vs [{Non-[City-State]} {Capital}] cities
This is definitely doable, we can add an IFilterable interface for the common functionality between different filter types

And this is something sorely lacking in tile filters, for example

I"m not planning on implementing OR conditionals due to confusion ("A and B or C" is ambiguous, "A and (B or C)" is too formal-logic-y) - and the last thing we need is to start implementing an AST in Video Game...

Also I wish there was a simple *readable* way to invert conditionals... but I haven't been able to think of one

**Comment by yairm210** [comment]
```kotlin

interface IFilterable {
    fun matchesFilter(filter: String): Boolean {
        if (filter.startsWith("Non-[") && filter.endsWith("]"))
            return !matchesFilter(filter.removePrefix("Non-[").removeSuffix("]"))
        if (filter.startsWith("{")) return filter.filterCompositeLogic({matchesFilter(it)}) { a, b -> a && b } ?: false

        return matchesFilterInternal(filter)
    }

    fun matchesFilterInternal(filter: String): Boolean
}

```

**Comment by yairm210** [comment]
That's what I mean by "readable"

"isn't before discovering tech" is not gramatically correct

**Comment by SpacedOutChicken** [comment]
We could just try <not> for it. 
`<not> <while the empire is happy>
<not> <on water maps>
<not> <when defending>`
etc.

It's not perfect, but it might be good enough.

**Comment by hackedpassword** [comment]
I went back to the G&K json's, applied my suggestion, immediately see the shortcomings as noted not being grammatically correct. Tenses and context make this a very muddy task. I called in backup on this one:
https://chat.openai.com/share/9f094db8-8cc7-42f4-902b-90543f1930ca

The proposed solution under Universal Prefixes is feasible. Rather than attempting to compensate for the many lexical forms required, a conditional trigger like these could be the best and simplest route to take:

```
"Does not apply: Provides the cheapest [Culture] building in your first [4] cities for free."
"Negates: [+15]% Production when constructing [All] wonders [in all cities]."
"Does not apply: Free [Worker] appears."
```

I'd imagine such triggers as:
<does not apply>
or
<negates>

vs <not> which is a bit pointy and unfriendly. (if this was just for myself and not a community I'd go with <not>)
